File: Python/530. Minimum Absolute Difference in BST.py
'''
1.递归
2.迭代
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
#         res = float('inf')
#         pre = None
#         def traversal(node):
#             nonlocal res, pre
#             if not node:
#                 return
#             if node.left:
#                 traversal(node.left)
#             if pre:
#                 res = min(res, node.val - pre.val)
#             pre = node
#             if node.right:
#                 traversal(node.right)
#         traversal(root)
#         return res

class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        st = [root]
        res = float('inf')
        pre = None
        while st:
            cur = st.pop(-1)
            if cur:
                if cur.right:
                    st.append(cur.right)
                st.append(cur)
                st.append(None)
                if cur.left:
                    st.append(cur.left)
            else:
                cur = st.pop(-1)
                if pre:
                    res = min(res, cur.val - pre.val)
                pre = cur
        return res

File: C++/530. Minimum Absolute Difference in BST.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 // class Solution {
 // public:
 //     int res = INT_MAX;
 //     TreeNode* pre = NULL;
 //     void traversal(TreeNode* root){
 //         if (!root) return;
 //         traversal(root->left);
 //         if (pre) res = min(res, root->val - pre->val);
 //         pre = root;
 //         traversal(root->right);
 //     }
 //     int getMinimumDifference(TreeNode* root) {
 //         traversal(root);
 //         return res;
 //     }
 // };

class Solution {
public:
    int res = INT_MAX;
    TreeNode* pre = NULL;
    int getMinimumDifference(TreeNode* root) {
        stack<TreeNode*> st;
        st.push(root);
        while (!st.empty()) {
            TreeNode* cur = st.top(); st.pop();
            if (cur != NULL) {
                if (cur->right) st.push(cur->right);
                st.push(cur);
                st.push(NULL);
                if (cur->left) st.push(cur->left);
            }
            else {
                TreeNode* cur = st.top(); st.pop();
                if (pre) res = min(res, cur->val - pre->val);
                pre = cur;
            }
        }
        return res;
    }
};